lib/src/main/scala/spinal/lib/bus/amba3/ahblite/AhbLite3Arbiter.scala

```diff
@@ -53,42 +72,72 @@ case class AhbLite3Arbiter(ahbLite3Config: AhbLite3Config, inputsCount: Int, rou
     val maskProposal    = Bits(inputsCount bits)
     val maskLocked      = Reg(Bits(inputsCount bits)) init(BigInt(1) << (inputsCount - 1))
     val maskRouted      = Mux(locked || dataPhaseActive, maskLocked, maskProposal)
+    val requestIndex    = OHToUInt(maskRouted)
+
+    /** Backup address phase */
+    val addressPhaseData  = Vec(Reg(AhbLite3AddrPhase(ahbLite3Config)), inputsCount)
+    val addressPhaseValid = Vec(RegInit(False), inputsCount)
+
+
+    for((input, data, valid) <- (io.inputs, addressPhaseData, addressPhaseValid).zipped){
+      when(input.HSEL & input.HTRANS === 2 & input.HREADYOUT){
+        data.assignFromBus(input)
+        valid := True
+      }
+    }
+
+    val transactionOnHold = addressPhaseValid.reduce(_ || _)
+
 
-    when(io.output.HSEL) { //valid
+    when(io.output.HSEL & io.output.HTRANS(1)) { // valid transaction
       maskLocked := maskRouted
       locked     := True
+      addressPhaseValid(requestIndex) := False
 
-      when(io.output.HREADY){ //fire
-        when(io.output.last && !io.output.HMASTLOCK) { //End of burst and no lock
-          locked := False
-        }
+      when(io.output.HREADY & io.output.last && !io.output.HMASTLOCK) { // End of burst and no lock
+        locked := False
```

Comment:
I very much dislike this code, both in its old and in its new form. [AMBA 3 AHB-Lite specification,](https://static.docs.arm.com/ihi0033/a/IHI0033A_new_eula.pdf) Section 3.5.2 states that

> Although masters are not permitted to terminate a burst request early, slaves must be designed to work correctly if the burst is not completed.

This means that decoding the `last` information is quite worthless and leading to wrong conclusions in unlocking from a burst. IMHO, the only condition for unlocking is checking for `HMASTLOCK`, and whether `HTRANS` is either `IDLE` or `NONSEQ`. I am aware of the implications on timing, but relying on `last` is improper.